File: 01-Sorting-Basic/Code-Python/insertion_sort_advance/insertion_sort.py
"""
优化后插入排序算法（升序）
"""


def sort(arr):
    for i in range(1, len(arr)):
        e = arr[i]
        j = i
        while j > 0 and e < arr[j - 1]:
            arr[j] = arr[j - 1]
            j -= 1
        arr[j] = e

File: codeJava/src/me.imtt/sorting/basic/InsertionSort.java
package me.imtt.sorting.basic;

/**
 * 插入排序（升序）
 */
public class InsertionSort {
    public static void sort(Comparable[] arr) {
        for (int i = 1; i < arr.length; i++) {
            for (int j = i; j > 0 && arr[j].compareTo(arr[j - 1]) < 0; j--) {
                swap(arr, j, j - 1);
            }
        }
    }

    /**
     * 交换数组中两个数
     */
    private static void swap(Object[] arr, int index1, int index2) {
        Object temp = arr[index1];
        arr[index1] = arr[index2];
        arr[index2] = temp;
    }
}
